horovod/data/data_loader_base.py

```diff
@@ -75,34 +76,57 @@ def close_async_loader(self):
         """
         Close the async data loader.
         """
-        print("Closing the AsyncDataLoaderMixin.")
+        print(f"close_async_loader[{self.async_loader_queue_size}], Closing the AsyncDataLoaderMixin.")
         if self.async_loader_queue_size > 0 and self.started:
             self.finished_event.set()
+            c = 0
             while True:
                 try:
                     # Drain buffer
                     self.queue.get_nowait()
+                    if self.debug_data_loader:
+                        print(f"close_async_loader[{self.async_loader_queue_size}], discarded batch #{c} from Queue.")
+
+                    c += 1
+
+                    # Force break out if hanging. We assume hanging if already pop enough items from queue.
+                    if c > max(2 * self.async_loader_queue_size, 200):
+                        print(f"close_async_loader: ERROR!!! Force break out after {c} times get_nowait.")
+                        break
                 except Empty:
                     break
+            print(f"close_async_loader[{self.async_loader_queue_size}], joining...")
             self.thread.join()
+            print(f"close_async_loader[{self.async_loader_queue_size}], joined.")
 
     def _async_worker(self):
         """
         Start worker thread to load data asynchronously.
         User need to implement self._iterate() to read the data.
         """
         try:
+            c = 0
             while not self.finished_event.is_set():
                 for batch in self._iterate():
                     if self.finished_event.is_set():
                         break
                     self.queue.put(batch)
+
+                    if self.debug_data_loader:
+                        print(f"_async_worker[{self.async_loader_queue_size}], push batch #{c}.")
+                        c += 1
+
+                if self.debug_data_loader:
+                    print(f"_async_worker[{self.async_loader_queue_size}], finish reading at #{c}, reset debugging counter, append None to queue.")
+                    c = 0
+
                 self.queue.put(None)
         except Exception as ex:
             self.queue.put(ex)
             self.queue.put(None)
         finally:
             self.queue.put(None)
+        print(f"_async_worker[{self.async_loader_queue_size}], stoped")
```

Comment:
put this print under `debug_data_loader`

Reply:
Like to keep this outside, it is only 1 time printing in the end. We saw some data loader hang issue during close, it  helps to detect that.